Automattic/crowdsignal-ui
# Switch to externally provided Gutenberg bundles for @crowdsignal/dashboard

This patch switches `@crowdsignal/dashboard`'s build process to use externally provided Gutenberg bundles which will ease integration with the isolated block editor and make upgrading different dependencies much easier.  
With this change come some pretty significant changes for our development and eventually production environment:

**Gutenberg is now installed in Crowdsignal UI as a git submodule.** This allows us to compile Gutenberg from source and ensure we're always using a version that actually works for us.  
Whenever we decide to update the Gutenberg version, the one that's actually available for builds using externals (currently just `@crowdsignal/dashboard`, all you need to do is `cd` into `lib/gutenberg`, checkout the correct tag/branch, and commit the change.

This also means **you now need to run `yarn build-gutenberg` once when initializing the repo, applying this patch or anytime there's been an update to the Gutenberg version being used**.

**Quirks:**
- External WordPress libraries need to be requested in the correct order respecting their dependencies on each other. I created a script in `scripts/resolve-wp-dependencies.php` that will generate a correct list based on the `*.asset.php` file generated for the app during the build.
- There are two reasons why the submodule approach is interesting: one is for the script described above to work, and the other is it will keep you sane while working with developer tools open as otherwise you'd be re-downloading all of these from the CDN on every single reload. It quickly gets tiring.

**Known issues**:
- Due to styling issues not yet addressed in Automattic/isolated-block-editor, I've currently locked us in at `WordPress/gutenberg@12.9.0`.
- The only `@wordpress/*` dependency that's still included in the bundle and cannot be provided externally is `@wordpress/interface`. This is pinned at `4.4.0` in `Automattic/isolated-block-editor` and needs to be pinned to the same version in any package depending on `Automattic/isolated-block-editor` in order to prevent there being multiple instances of the same code running causing issues.
- All other `@wordpress/` packages should still be specified to use the latest compatible version (`^` prefix). We still have packages that compile into a single bundle (with minimal `@wordpress/*` dependencies but still...) and we want to prevent issues with duplicated dependencies there as well. It doesn't matter if `@crowdsignal/dashboard` ends up with duplicates (other than aforementioned `@wordpress/interface`) as these are provided externally anyway.

# Testing

Needs D80226-code.

- Run `yarn build-gutenberg`. This should install the git submodule, install Gutenberg's dependencies and build it.
- Run `yarn start`.

Hopefully, at this point you see no difference to before & after applying the patch 😂 
If you look deeper, you should see a bunch of libraries load from `/gutenberg` and our production bundle size will be reduced.
